Add tests for BSBLAN climate component

liudger opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
Make the test coverage 100%

## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

liudger committed:
chore: Add tests for BSBLAN climate component

liudger committed:
fix return types

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,186 @@
+    state.current_temperature = MockStateValue("22.5")
+    state.target_temperature = MockStateValue("21.0")
+    state.hvac_mode = MockStateValue("heat")
+    return state
```
Rather just move stuff to conftest and make a mock you use for the whole integration tests instead of climate specifically

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,186 @@
+    return state
+
+
+@pytest.fixture
+def coordinator_mock(mock_state) -> MagicMock:
+    """Create a mock BSBLan coordinator."""
+    coordinator = AsyncMock()
+    coordinator.data = BSBLanCoordinatorData(state=mock_state())
+    return coordinator
```
Don't create classes in the tests. We want to set up the integration like it does in real life and then observe the state machine

liudger committed:
fix MAC data

liudger committed:
chore: Update BSBLAN climate component tests

used setup from conftest  added setup for farhenheit temp unit

joostlek commented on tests/components/bsblan/conftest.py:
```diff
@@ -71,3 +77,44 @@ async def init_integration(
+
+@pytest.fixture
+def climate(hass: HomeAssistant, mock_config_entry, mock_bsblan) -> BSBLANClimate:
+    """Set up the BSBLan climate entity for testing."""
+    mock_device_registry(hass)
+    coordinator = BSBLanUpdateCoordinator(hass, mock_config_entry, mock_bsblan)
+    coordinator.config_entry = mock_config_entry
+    coordinator.data = BSBLanCoordinatorData(state=mock_bsblan.state.return_value)
+    data = BSBLanData(
+        client=mock_bsblan,
+        coordinator=coordinator,
+        device=mock_bsblan.device.return_value,
+        info=mock_bsblan.info.return_value,
+        static=mock_bsblan.static_values.return_value,
+    )
+    hass.data.setdefault("bsblan", {})[mock_config_entry.entry_id] = data
+    return BSBLANClimate(data)
+
+
+@pytest.fixture
+def climate_fahrenheit(
+    hass: HomeAssistant, mock_config_entry, mock_bsblan
+) -> BSBLANClimate:
+    """Set up the BSBLan climate entity with Fahrenheit temperature unit."""
+    mock_device_registry(hass)
+    coordinator = BSBLanUpdateCoordinator(hass, mock_config_entry, mock_bsblan)
+    coordinator.config_entry = mock_config_entry
+    coordinator.data = BSBLanCoordinatorData(state=mock_bsblan.state.return_value)
+    # override min_temp unit to Fahrenheit in mock_bsblan static values
+    mock_bsblan.static_values.return_value.min_temp.unit = "°F"
+
+    data = BSBLanData(
+        client=mock_bsblan,
+        coordinator=coordinator,
+        device=mock_bsblan.device.return_value,
+        info=mock_bsblan.info.return_value,
+        static=mock_bsblan.static_values.return_value,
+    )
+    hass.data.setdefault("bsblan", {})[mock_config_entry.entry_id] = data
+    return BSBLANClimate(data)
```
This is still not what we want. Please check an integration like `airgradient` and `flexit_bacnet` for a good example. We want the integration to act like a real life example

liudger replied:
not yet 100% now. I'll have another look how I can increase the coverage with this method.

joostlek replied:
Check the air gradient tests. we can setup the integration, call a service and then see that the library is called in a right way

liudger committed:
chore: Update BSBLAN climate component tests

use syrupy to compare results

liudger committed:
add test for temp_unit

liudger committed:
update climate tests

set current_temperature to None in test case. Is this the correct way for testing?

liudger commented:
@joostlek to get all attributes tested I need to set some values that bsblan could return depending on the features. Is this the way? [a9ec445](https://github.com/home-assistant/core/pull/124524/commits/a9ec4451bb8b70ca96112b534d7ecefe49057bbd) It was hard to get it from the other examples how to fake some returns to let the integration return the correct values

joostlek commented:
> get all attributes tested 

Which ones?
Also there is a merge conflict

liudger commented:
the current_temperature is None
```
# Spoof the current_temperature value to "---"
    mock_bsblan.set_current_temperature("---")

    # Update the state in Home Assistant
    await hass.helpers.entity_component.async_update_entity("climate.bsb_lan")

    # Get the state of the climate entity
    state = hass.states.get("climate.bsb_lan")

    # Assert that the current_temperature attribute is None
    assert state.attributes["current_temperature"] is None
    ```
    
    Yeah I will fix the merge conflict. Accidentally fixed a bug twice.

joostlek commented:
But you can repatch the object to return "---" righjt

liudger committed:
chore: Update BSBLAN diagnostics to handle asynchronous data retrieval

liudger committed:
chore: Refactor BSBLAN conftest.py to simplify fixture and patching

liudger committed:
chore: Update BSBLAN climate component tests

100% test coverage

liudger committed:
chore: Update BSBLAN diagnostics to handle asynchronous data retrieval

liudger committed:
chore: Update snapshots

liudger committed:
Fix BSBLAN climate test for async_set_preset_mode

- Update test_async_set_preset_mode to correctly handle ServiceValidationError
- Check for specific translation key instead of full error message
- Ensure consistency between local tests and CI environment
- Import ServiceValidationError explicitly for clarity

joostlek commented on homeassistant/components/bsblan/diagnostics.py:
```diff
@@ -18,10 +19,15 @@ async def async_get_config_entry_diagnostics(
     return {
-        "info": data.info.to_dict(),
-        "device": data.device.to_dict(),
+        "info": await data.info.to_dict()
```
why would `to_dict` be a coroutine?

joostlek commented on homeassistant/components/bsblan/entity.py:
```diff
@@ -22,7 +22,7 @@ class BSBLanEntity(CoordinatorEntity[BSBLanUpdateCoordinator]):
         """Initialize BSBLan entity."""
         super().__init__(coordinator, data)
-        host = coordinator.config_entry.data["host"]
+        host = self.coordinator.config_entry.data["host"]
```
```suggestion
        host = coordinator.config_entry.data["host"]
```

joostlek commented on tests/components/bsblan/conftest.py:
```diff
@@ -53,9 +53,13 @@ def mock_bsblan() -> Generator[MagicMock]:
-        bsblan.static_values.return_value = StaticState.from_json(
-            load_fixture("static.json", DOMAIN)
-        )
+        # Patch static values based on the test case
+        async def set_static_values(param):
+            bsblan.static_values.return_value = StaticState.from_json(
+                load_fixture(param, DOMAIN)
+            )
+
+        bsblan.set_static_values = set_static_values
```
Let's not do this here, we should make the mock be like the real life object ideally

joostlek replied:
Why don't we just set the static_values to a good default and then just overwrite in the test when we need to?

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+    await setup_with_selected_platforms(hass, mock_config_entry, [Platform.CLIMATE])
+    await snapshot_platform(hass, entity_registry, snapshot, mock_config_entry.entry_id)
+
+    climate_entity = hass.data[CLIMATE_DOMAIN].get_entity(ENTITY_ID)
```
Don't touch internals

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+    climate_entity = hass.data[CLIMATE_DOMAIN].get_entity(ENTITY_ID)
+
+    # Test when current_temperature is "---"
+    with patch.object(
+        mock_bsblan.state.return_value.current_temperature, "value", "---"
+    ):
```
```suggestion
    mock_bsblan.state.return_value.current_temperature = "---"
```

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+    with patch.object(
+        mock_bsblan.state.return_value.current_temperature, "value", "---"
+    ):
+        await hass.helpers.entity_component.async_update_entity(ENTITY_ID)
+        state = hass.states.get(ENTITY_ID)
```
Make the entity update naturally, so instead of doing this (I think this is a polling integration)) use the freezer and skip the same amount of time until the next update. It will then naturally update the entity

liudger replied:
I'll check the freezer

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+        assert state.attributes["temperature"] == target_temperature
+
+    # Test hvac_mode when preset_mode is ECO
+    with patch.object(mock_bsblan.state.return_value.hvac_mode, "value", PRESET_ECO):
+        assert climate_entity.hvac_mode == HVACMode.AUTO
+
+    # Test hvac_mode with other values
+    with patch.object(mock_bsblan.state.return_value.hvac_mode, "value", HVACMode.HEAT):
+        assert climate_entity.hvac_mode == HVACMode.HEAT
```
we should check this in the state machine, not directly at the object

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+    await mock_bsblan.set_static_values(static_file)
+    await setup_with_selected_platforms(hass, mock_config_entry, [Platform.CLIMATE])
+
+    climate_entity = hass.data[CLIMATE_DOMAIN].get_entity(ENTITY_ID)
+
+    with patch.object(climate_entity, "async_set_data") as mock_set_data:
+        await climate_entity.async_set_hvac_mode(mode)
+        mock_set_data.assert_called_once_with(hvac_mode=expected_call)
```
We should use service calls to change the hvac mode and then you can use your `mock_bsblan` to check if it did the right thing

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,299 @@
+    await mock_bsblan.set_static_values(static_file)
+    await setup_with_selected_platforms(hass, mock_config_entry, [Platform.CLIMATE])
+
+    climate_entity = hass.data[CLIMATE_DOMAIN].get_entity(ENTITY_ID)
+
+    with patch(
+        "homeassistant.components.bsblan.climate.BSBLANClimate.hvac_mode",
+        new_callable=PropertyMock,
+    ) as mock_hvac_mode:
+        mock_hvac_mode.return_value = hvac_mode
+
+        if expected_call is None:
+            with pytest.raises(ServiceValidationError) as exc_info:
+                await climate_entity.async_set_preset_mode(preset_mode)
+            assert exc_info.value.translation_key == "set_preset_mode_error"
+        else:
+            with patch.object(climate_entity, "async_set_data") as mock_set_data:
+                await climate_entity.async_set_preset_mode(preset_mode)
+                mock_set_data.assert_called_once_with(preset_mode=expected_call)
```
idem

liudger committed:
Update homeassistant/components/bsblan/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

liudger committed:
chore: Update BSBLAN conftest.py to simplify fixture and patching

liudger committed:
chore: Update BSBLAN integration setup function parameter name

liudger committed:
chore: removed set_static_value

liudger committed:
refactor: Improve BSBLANClimate async_set_preset_mode method

This commit refactors the async_set_preset_mode method in the BSBLANClimate class to improve code readability and maintainability. The method now checks if the HVAC mode is not set to AUTO and the preset mode is not NONE before raising a ServiceValidationError.

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

liudger committed:
refactor: Improve tests

test_celsius_fahrenheit
test_climate_entity_properties
test_async_set_hvac_mode

test_async_set_preset_mode still broken. Not sure why hvac mode will not set. THis causes error with preset mode set

liudger committed:
update snapshot

liudger committed:
fix DOMAIN bsblan

liudger committed:
refactor: Improve BSBLANClimate async_set_data method

liudger committed:
refactor: fix last tests

liudger committed:
refactor: Simplify async_get_config_entry_diagnostics method

liudger committed:
refactor: Improve BSBLANClimate async_set_temperature method

This commit improves the async_set_temperature method in the BSBLANClimate class. It removes the unnecessary parameter "expected_result" and simplifies the code by directly calling the service to set the temperature. The method now correctly asserts that the thermostat method is called with the correct temperature.

liudger committed:
refactor: Add static data to async_get_config_entry_diagnostics

liudger committed:
refactor: Add static data to async_get_config_entry_diagnostics

right place

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,311 @@
+
+    # Attempt to set the preset mode
+    if not expected_success:
+        with pytest.raises(HomeAssistantError) as exc_info:
+            await hass.services.async_call(
+                CLIMATE_DOMAIN,
+                SERVICE_SET_PRESET_MODE,
+                {ATTR_ENTITY_ID: ENTITY_ID, ATTR_PRESET_MODE: preset_mode},
+                blocking=True,
+            )
+        assert "set_preset_mode_error" in str(exc_info.value)
```
```suggestion
        with pytest.raises(HomeAssistantError, matches="Can't set preset mode to eco when HVAC mode is not set to auto"):
            await hass.services.async_call(
                CLIMATE_DOMAIN,
                SERVICE_SET_PRESET_MODE,
                {ATTR_ENTITY_ID: ENTITY_ID, ATTR_PRESET_MODE: preset_mode},
                blocking=True,
            )
```

liudger replied:
This gives an error is should be match? Also this won't work as there is more in there error message than the message.

liudger replied:
mine mistake. The error message is somehow deleted on the set preset mode. Now I fixed it and match does work

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,311 @@
+    mock_bsblan.state.return_value.hvac_mode = mock_hvac_mode
+
+    # Attempt to set the preset mode
+    if not expected_success:
```
Don't have if statements in the code

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,311 @@
+            {ATTR_ENTITY_ID: ENTITY_ID, ATTR_PRESET_MODE: preset_mode},
+            blocking=True,
+        )
+        await hass.async_block_till_done()
+
+    # Reset the mock for the next iteration
+    mock_bsblan.thermostat.reset_mock()
```
```suggestion
```
Unless you have some fixture with `scope=module`, the mock is reset everytime

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,311 @@
+
+    # Test error handling
+    mock_bsblan.thermostat.side_effect = BSBLANError("Test error")
+    with pytest.raises(HomeAssistantError) as exc_info:
+        await hass.services.async_call(
+            CLIMATE_DOMAIN,
+            SERVICE_SET_TEMPERATURE,
+            {ATTR_ENTITY_ID: ENTITY_ID, ATTR_TEMPERATURE: 20},
+            blocking=True,
+        )
+    assert "An error occurred while updating the BSBLAN device" in str(exc_info.value)
+    assert exc_info.value.translation_key == "set_data_error"
```
Please use matches= here as well

joostlek commented on tests/components/bsblan/test_climate.py:
```diff
@@ -0,0 +1,311 @@
+    mode: HVACMode,
+) -> None:
+    """Test setting HVAC mode via service call."""
+    static_data = json.loads(load_fixture(static_file, DOMAIN))
+    with patch.object(
+        mock_bsblan, "static_values", return_value=StaticState.from_dict(static_data)
+    ):
```
```suggestion
    mock_bsblan.static_values.return_value = StaticState.from_dict(load_json_object_fixture(static_file, DOMAIN))
```

joostlek reviewed (changes requested):
Oh and they are failing

liudger committed:
refactor: Improve error message for setting preset mode

This commit updates the error message in the BSBLANClimate class when trying to set the preset mode.

liudger committed:
refactor: Improve tests

liudger commented:
@joostlek is it possible to get this in 2024.9.2 release? The 2024.9.1  release bsblan is still broken. Tested this version well and seems to be working fine.

joostlek committed:
Fix
